### PR TITLE
[Feature/auth/password reset] 비밀번호 재설정 api 구현 및 테스트 작성

### PR DESCRIPTION
**Pull request by ds8600**
## ✅ PR 요약
- 이메일 인증 코드를 기반으로 비밀번호를 재설정할 수 있는 `POST /api/v1/auth/password/reset/` 엔드포인트를 추가했습니다.

## 📄 상세 내용
-  `PasswordResetAPIView`와 `auth-password-reset` 라우트를 추가하고 요청/응답 스키마를 정의했습니다.
- `auth_service`에 비밀번호 재설정 로직을 분리해 인증 코드 검증, 비밀번호 유효성 검사, 소셜 전용 계정 예외 처리를 구현했습니다.
-  성공, 잘못된 코드, 만료 코드, 소셜 계정, 비밀번호 형식 오류에 대한 테스트를 추가했습니다.

## 🧪 테스트
- 로컬에서 확인했습니다.

### CODE REVIEW

File: apps/users/services/auth_service.py [hunk 1]
```diff
@@ -86,6 +86,28 @@ def authenticate_user(*, email: str, password: str) -> User:
     return user
 
 
+def reset_user_password(*, email: str, code: str, new_password: str) -> None:
+    saved_code = cache.get(f"email_code:{email}")
+    if saved_code is None or not secrets.compare_digest(str(saved_code), code):
+        raise CustomAPIException(ErrorMessages.INVALID_CODE)
```
Comment (jhjdl3984) [inline apps/users/services/auth_service.py:133]: signup_user처럼 나눠서 CODE_EXPIRED로 받는게 좋을거같아요!

File: apps/users/services/auth_service.py [hunk 2]
```diff
@@ -86,6 +86,28 @@ def authenticate_user(*, email: str, password: str) -> User:
...
+    if user.social_accounts.exists() and not user.has_usable_password():
+        raise CustomAPIException(ErrorMessages.SOCIAL_USER_ONLY)
+
+    try:
+        validate_password(new_password)
+    except DjangoValidationError as err:
+        raise CustomAPIException(ErrorMessages.VALIDATION_ERROR) from err
+
+    user.set_password(new_password)
+    user.save(update_fields=["password"])
```
Comment (jhjdl3984) [inline apps/users/services/auth_service.py:148]: 비밀번호 변경되면 기존 로그인된 상태를 강제 로그아웃 하는 처리 완료 후 기존 세션 전체 무효화가 없는거 같아요!

File: apps/users/services/auth_service.py [hunk 3]
```diff
@@ -86,6 +86,28 @@ def authenticate_user(*, email: str, password: str) -> User:
...
+    cache.delete(f"email_code:{email}")
```
Comment (ji-min0) [inline apps/users/services/auth_service.py]: 인증 코드 검증 실패 시 시도 횟수 제한이 없는 것 같습니다.

현재 구조에서는 동일 email에 대해 code 검증을 무제한으로 시도할 수 있어
6자리 코드 특성상 brute force 공격 가능성이 있습니다.

아래와 같은 제한을 두는 것도 좋을 것 같습니다.
- code 검증 시도 횟수 제한 (예: 5회)
- 실패 횟수 초과 시 code 무효화
- 또는 일정 시간 동안 재시도 차단

이 부분 보안 측면에서 고려가 필요할 것 같습니다.

File: apps/users/services/auth_service.py
Comment (ji-min0) [inline apps/users/services/auth_service.py]: 현재 이메일 인증 코드 로직이 send_signup_email_code로 구현되어 있는데
실제로는 회원가입뿐 아니라 비밀번호 재설정에서도 동일한 code를 사용하고 있는 구조로 보입니다.

`cache.set(f"email_code:{email}", code, timeout=EMAIL_CODE_TTL_SECONDS)
`
현재 구조에서는 email code의 purpose가 분리되어 있지 않아서 회원가입 코드와 비밀번호 재설정 코드 가 동일하게 사용될 수 있는 상태입니다.

이 경우 인증 도메인 의미가 혼재될 수 있어 아래와 같이 purpose를 분리해서 관리해도 좋을 것 같습니다. 

`email_code:{purpose}:{email}
` 

(예)
- email_code:signup:{email}
- email_code:password_reset:{email}

또는 send_email_code API에 purpose 파라미터를 두는 방식도 고려할 수 있을 것 같습니다.

현재 send_signup_email_code라는 함수 이름과 실제 사용 범위도 약간 불일치하는 부분이 있어 인증 코드 로직을 purpose 기반으로 분리하는 것도 검토해보면 좋을 것 같습니다

File: apps/users/services/auth_service.py [hunk 4]
```diff
@@ -86,6 +87,35 @@ def authenticate_user(*, email: str, password: str) -> User:
...
+
+    try:
+        validate_password(new_password)
+    except DjangoValidationError as err:
+        raise CustomAPIException(ErrorMessages.VALIDATION_ERROR) from err
+
+    user.set_password(new_password)
+    user.save(update_fields=["password"])
+    revoke_all_refresh_tokens(user)
+    cache.delete(f"email_code:{email}")
```
Comment (ji-min0) [inline apps/users/services/auth_service.py]: reset_user_password 내부에서는 기존 로그인 세션을 무효화하는 로직이 보이지 않는 것 같습니다.

```
user.set_password(new_password)
user.save(update_fields=["password"])
```
현재 구조에서는 기존 access token / refresh token이 그대로 유효할 가능성이 있습니다.

비밀번호 변경 시 일반적으로 다음과 같은 처리를 합니다.

- 기존 refresh token blacklist 처리
- 사용자 token version 증가
- 또는 전체 세션 invalidation 처리

이 부분 확인이 필요해 보입니다.

File: apps/users/services/auth_service.py [hunk 5]
```diff
@@ -86,6 +87,35 @@ def authenticate_user(*, email: str, password: str) -> User:
     return user
```
Comment (ji-min0) [inline apps/users/services/auth_service.py]:
```
if User.objects.filter(email=email).exists():
        raise CustomAPIException(ErrorMessages.EMAIL_ALREADY_EXISTS)
```
현재 /auth/email/code API를 통해 이메일 가입 여부를 확인할 수 있는 구조로 보입니다.

이 경우 공격자가 이메일 리스트를 기반으로 회원 존재 여부를 수집할 수 있는 User Enumeration 문제가 발생할 수 있습니다.

보안상 아래와 같은 방법을 생각해 볼 수 있을 것 같습니다.
- 이메일 존재 여부와 관계없이 동일 응답 반환
- "인증 코드가 발송되었습니다" 메시지 고정
- 실제 코드 발송 여부만 내부에서 처리

또는 회원가입 / 비밀번호 재설정 이메일 코드 API를 분리하는 방식도 고려할 수 있을 것 같습니다.

File: apps/users/tests/test_password_reset.py
Comment (ji-min0) [inline apps/users/tests/test_password_reset.py:1]: 추가로 아래 케이스도 있으면 좋을 것 같습니다.

- 동일 code 재사용 시 실패 케이스
- 여러 번 잘못된 code 시도 케이스
- 존재하지 않는 email 요청 케이스

인증 관련 로직은 edge case 테스트가 많을수록 안정성이 높아질 것 같습니다.

File: apps/users/services/auth_service.py (hunk 1 above)
Comment (ds8600) [reply apps/users/services/auth_service.py:133]: 보안 이슈!

**Review by ji-min0 — CHANGES_REQUESTED** [review]

**Review by jhjdl3984 — DISMISSED** [review]

File: apps/users/services/auth_service.py (hunk 3 above)
Comment (ds8600) [reply apps/users/services/auth_service.py]: 반영했습니다.
- 인증 코드 검증 실패는 5회 / 20분 제한을 두고, 초과 시 `TOO_MANY_REQUESTS`를 반환하도록 수정했습니다.

File: apps/users/services/auth_service.py
Comment (ds8600) [reply apps/users/services/auth_service.py]: - email code cache key를 `email_code:{purpose}:{email}` 형태로 분리했습니다.

File: apps/users/services/auth_service.py (hunk 4 above)
Comment (ds8600) [reply apps/users/services/auth_service.py]: - 비밀번호 재설정 성공 시 기존 refresh token은 전체 blacklist 처리하도록 보완했습니다.

File: apps/users/services/auth_service.py (hunk 5 above)
Comment (ds8600) [reply apps/users/services/auth_service.py]: - `/auth/email/code`는 이메일 존재 여부와 관계없이 동일 응답을 반환하도록 변경했습니다.

File: apps/users/tests/test_password_reset.py
Comment (ds8600) [reply apps/users/tests/test_password_reset.py:1]: - 동일 code 재사용, 잘못된 code 반복 시도, 존재하지 않는 email 등의 edge case 테스트도 추가했습니다.

**Review by jhjdl3984 — APPROVED** [review]

**Review by ji-min0 — APPROVED** [review]
